src/keygen.rs

```diff
@@ -0,0 +1,171 @@
+    ),
+)]
+/// Runs a distributed key generation (DKG) process using DFNS-CGGMP21 protocol
+///
+/// # Arguments
+/// * `n` - Number of parties participating in the DKG
+/// * `context` - The DFNS context containing network and storage configuration
+///
+/// # Returns
+/// Returns the generated public key as a byte vector on success
+///
+/// # Errors
+/// Returns an error if:
+/// - Failed to retrieve blueprint ID or call ID
+/// - Failed to get party information
+/// - MPC protocol execution failed
+/// - Serialization of results failed
+pub async fn keygen(n: u16, context: DfnsContext) -> Result<Vec<u8>, GadgetError> {
+    // Get configuration and compute deterministic values
+    let blueprint_id = context
+        .blueprint_id()
+        .map_err(|e| KeygenError::ContextError(e.to_string()))?;
+    let call_id = context
+        .current_call_id()
+        .await
+        .map_err(|e| KeygenError::ContextError(e.to_string()))?;
+
+    let (meta_hash, deterministic_hash) = compute_deterministic_hashes(n, blueprint_id, call_id);
+    let execution_id = ExecutionId::new(&deterministic_hash);
+
+    // Setup party information
+    let (party_index, operators) = context
+        .get_party_index_and_operators()
+        .await
+        .map_err(|e| KeygenError::ContextError(e.to_string()))?;
+
+    let parties: BTreeMap<u16, Public> = operators
+        .into_iter()
+        .enumerate()
+        .map(|(j, (_, ecdsa))| (j as u16, ecdsa))
+        .collect();
+
+    gadget_sdk::info!(
+        "Starting DFNS-CGGMP21 Keygen for party {party_index}, n={n}, eid={}",
+        hex::encode(execution_id.as_bytes())
+    );
+
+    // Initialize RNG and network
+    let mut rng = ChaCha20Rng::from_seed(deterministic_hash);
+    let delivery = setup_network_party(&context, party_index, deterministic_hash, parties).await;
+    let party = MpcParty::connected(delivery);
+    // Execute the MPC protocol
+    let result = KeygenBuilder::<Secp256k1, SecurityLevel128, Sha256>::new(
+        execution_id,
+        party_index as u16,
+        n,
+    )
+    .start(&mut rng, party)
+    .await
+    .map_err(|e| KeygenError::MpcError(e.to_string()))?;
+
+    gadget_sdk::info!(
+        "Ending DFNS-CGGMP21 Keygen for party {party_index}, n={n}, eid={}",
+        hex::encode(execution_id.as_bytes())
+    );
+
+    // Store the results
+    let store_key = hex::encode(meta_hash);
+    context.store.set(
+        &store_key,
+        DfnsStore {
+            inner: Some(result.clone()),
+            refreshed_key: None,
+        },
+    );
+
+    // Serialize the results
+    let public_key = bincode::serialize(&result.shared_public_key)
+        .map_err(|e| KeygenError::SerializationError(e.to_string()))?;
+
+    // Serialize the share (currently unused but kept for potential future use)
+    let _serializable_share = bincode::serialize(&result.into_inner())
+        .map_err(|e| KeygenError::SerializationError(e.to_string()))?;
```

Comment:
We do not recommend using bincode for share serialization. We expanded on this in the [docs](https://docs.rs/cggmp21/latest/cggmp21/key_share/struct.DirtyIncompleteKeyShare.html#serialization-format-via-serde)